Source/loadsave.cpp

```diff
@@ -1,700 +1,1644 @@
+		for (i = 0; i < MAXMISSILES; i++)
+			missileavail[i] = BLoad();
+		for (i = 0; i < nummissiles; i++)
+			LoadMissile(missileactive[i]);
+		for (i = 0; i < MAXOBJECTS; i++)
+			objectactive[i] = BLoad();
+		for (i = 0; i < MAXOBJECTS; i++)
+			objectavail[i] = BLoad();
+		for (i = 0; i < nobjects; i++)
+			LoadObject(objectactive[i]);
+		for (i = 0; i < nobjects; i++)
+			SyncObjectAnim(objectactive[i]);
+
+		numlights = WLoad();
+
+		for (i = 0; i < MAXLIGHTS; i++)
+			lightactive[i] = BLoad();
+		for (i = 0; i < numlights; i++)
+			LoadLighting(lightactive[i]);
+
+		visionid = WLoad();
+		numvision = WLoad();
+
+		for (i = 0; i < numvision; i++)
+			LoadVision(i);
+	}
+
+	for (i = 0; i < MAXITEMS; i++)
+		itemactive[i] = BLoad();
+	for (i = 0; i < MAXITEMS; i++)
+		itemavail[i] = BLoad();
+	for (i = 0; i < numitems; i++)
+		LoadItem(itemactive[i]);
+	for (i = 0; i < 128; i++)
+		UniqueItemFlag[i] = OLoad();
+
+	for (j = 0; j < MAXDUNY; j++) {
+		for (i = 0; i < MAXDUNX; i++)
+			dLight[i][j] = BLoad();
+	}
+	for (j = 0; j < MAXDUNY; j++) {
+		for (i = 0; i < MAXDUNX; i++)
+			dFlags[i][j] = BLoad();
+	}
+	for (j = 0; j < MAXDUNY; j++) {
+		for (i = 0; i < MAXDUNX; i++)
+			dPlayer[i][j] = BLoad();
+	}
+	for (j = 0; j < MAXDUNY; j++) {
+		for (i = 0; i < MAXDUNX; i++)
+			dItem[i][j] = BLoad();
+	}
+
+	if (leveltype != DTYPE_TOWN) {
+		for (j = 0; j < MAXDUNY; j++) {
+			for (i = 0; i < MAXDUNX; i++)
+				dMonster[i][j] = WLoad();
+		}
+		for (j = 0; j < MAXDUNY; j++) {
+			for (i = 0; i < MAXDUNX; i++)
+				dDead[i][j] = BLoad();
+		}
+		for (j = 0; j < MAXDUNY; j++) {
+			for (i = 0; i < MAXDUNX; i++)
+				dObject[i][j] = BLoad();
+		}
+		for (j = 0; j < MAXDUNY; j++) {
+			for (i = 0; i < MAXDUNX; i++)
+				dLight[i][j] = BLoad();
+		}
+		for (j = 0; j < MAXDUNY; j++) {
+			for (i = 0; i < MAXDUNX; i++)
+				dPreLight[i][j] = BLoad();
+		}
+		for (j = 0; j < DMAXY; j++) {
+			for (i = 0; i < DMAXX; i++)
+				automapview[i][j] = OLoad();
+		}
+		for (j = 0; j < MAXDUNY; j++) {
+			for (i = 0; i < MAXDUNX; i++)
+				dMissile[i][j] = BLoad();
+		}
+	}
+
+	numpremium = WLoad();
+	premiumlevel = WLoad();
+
+	for (i = 0; i < 6; i++)
+		LoadPremium(i);
+
+	automapflag = OLoad();
+	AutoMapScale = WLoad();
+	mem_free_dbg(LoadBuff);
+	AutomapZoomReset();
+	ResyncQuests();
+
+	if (leveltype != DTYPE_TOWN)
+		ProcessLightList();
+
+	RedoPlayerVision();
+	ProcessVisionList();
+	missiles_process_charge();
+	ResetPal();
+	SetCursor_(CURSOR_HAND);
+	gbProcessPlayers = TRUE;
+}
+
+char BLoad()
+{
+	return *tbuff++;
+}
+
+int WLoad()
+{
+	int rv = *tbuff++ << 24;
+	rv |= *tbuff++ << 16;
+	rv |= *tbuff++ << 8;
+	rv |= *tbuff++;
+
+	return rv;
+}
+
+int ILoad()
+{
+	int rv = *tbuff++ << 24;
+	rv |= *tbuff++ << 16;
+	rv |= *tbuff++ << 8;
+	rv |= *tbuff++;
+
+	return rv;
+}
+
+BOOL OLoad()
+{
+	if (*tbuff++ == TRUE)
+		return TRUE;
+	else
+		return FALSE;
+}
+
+void CopyInt(const void *src, void *dst)
+{
+	memcpy(dst, src, 4);
+	tbuff += 4;
```

Comment:
It feels a bit odd that the function takes `src` yet operates on `tbuff`, probably better to just use `tbuff` directly in both places. Also does it have any benifits over using the existing `ILoad()`?

Reply:
I was initially trying to use a single function to pack and unpack these structs, so I wrote Copy*() functions that can be used for both packing and unpacking, automatically incrementing tbuff. 
In the case of packing, the dst pointer (tbuff) would need to be incremented, and in the case of unpacking, the source pointer (tbuff) would need to be incremented. But since the function is more generic than ILoad(), it can't know which pointer to increment without passing it extra args. So this keeps a single function for packing and unpacking Ints from the MPQ save data while still keeping a light function signature.